### PR TITLE
use CrossVersion.full for unstable Scala by default

### PR DESCRIPTION
**Pull request by eed3si9n**
### sbt 0.12.1 behavior

```
> ++ 2.10.0-RC2
Setting version to 2.10.0-RC2
[info] Set current project to helloworld (in build file:/Users/foo/bar/)
> cross-version
[info] Binary
> scala-binary-version
[info] 2.10
```

The cross version is `Binary` by default. This publishes the application with `_2.10` suffix.
### what this changes

```
> ++ 2.10.0-RC2
Setting version to 2.10.0-RC2
[info] Set current project to default-c60480 (in build file:/Users/foo/bar/)
> cross-version
[info] Binary
> scala-binary-version
[info] 2.10.0-RC2
```

The cross version is still `Binary`, but the binary version is calculated to be `2.10.0-RC2` since it's non-final.
### historical notes

The originating commit for this behavior is harrah/xsbt@13e62fd6452974a77eb9b9ac9a60a613316b1d14 Use binary version for cross-version even for snapshots and milestones, and it is behaving as designed. The commit message reads:

> Rely instead on users not publishing the same stable version against both stable Scala or sbt releases and snapshots/milestones.

The change was applied both to the binary versions of sbt and Scala, and was first introduced in sbt [0.12.0-RC1](https://groups.google.com/d/topic/simple-build-tool/eAGOACDH59E/discussion), which allowed sbt 0.12 plugins to be published ahead of sbt 0.12.0 release.

Unlike sbt that's been keeping the compatibility, there's no such guarantee for Scala milestones and RCs. Here are related threads on ML:
- [Scala version suffix in published artifact names](https://groups.google.com/forum/#!topic/simple-build-tool/miiQwiuxuIY/discussion)
- [the price of the wrong defaults](https://groups.google.com/d/topic/simple-build-tool/d_Kd6H_QMmM/discussion)

In the former thread it was suggested:

>  I would use a snapshot, a version bump, or append the full version.

There are pressures on library authors to cross publish against 2.10.0 RCs driven by user demands and curiosity for new features added in 2.10, and most authors I think are opting for full version instead of snapshot or version bump. See [community projects published for RC2](https://groups.google.com/forum/#!topic/scala-language/rbUwCwFEiIE/discussion).

This is a matter of setting `crossVersion`, so each project can work around this by knowing about this issue and writing logic around `scalaVersion`. For example the following would replicate the sbt 0.12.0 beta behavior, which I assume many are using:

``` scala
crossVersion <<= scalaVersion { sv => if (sv contains "-") CrossVersion.full else CrossVersion.binary }
```
### opinion

Since not every sbt user is an sbt geek, the default behavior should be safer.

### CODE REVIEW

**Comment by jsuereth** [comment]
So a few technical notes:

`sv contains "-"` is _NOT_ enough to figure out if we're on an unstable release, it's far more complicated than that.

Specifically, look at these two version numbers, all stable against some binary candidate:
- `2.10.0-1` - Hotfix release.  Stable against `2.10`
- `2.10.1-RC1` - Binary Compatible bugfix RC against `2.10`

The second is the more worrisome of the two.   That is that with Scala, BC is locked to _minor releases_ which means new bugfix releases _are_ binary compatible.  The only instance where we have issues is when releasing  a new `X.Y.0` release, but not a `X.Y.0-N` release.   

This also doesn't set the best precedent for other projects to follow.

A full solution would be to use repositories that _embed BC information_ into the _artifact metadata_ (Why, hello ELF).   Barring that, we have hacky scripts like what's in SBT.  

I totally understand the goal of improving, but this one swings a bit too far, and again jars the community into having to relearn how to deal with binary compatibility.   

Do you think we can find an alternative practical solution for now?  I'm all ears to ideas.

**Comment by eed3si9n** [comment]
> `sv contains "-"` is NOT enough to figure out if we're on an unstable release, it's far more complicated than that.

That's how `isStable` is [defined now](https://github.com/harrah/xsbt/blob/5e155900da9b0ad549405bf0e1b0953006385e68/ivy/CrossVersion.scala#L70):

``` scala
def isStable(v: String): Boolean = !v.contains("-")
```

I agree with your point, but I didn't want to muddy the pull request with my attempt of approximating stability.
The reality is that we have sbt 0.12.x that's shipping jars for `_0.12`, and people are telling each other how to work around it like [this protip](https://twitter.com/sgodbillon/status/269825436400373760):

> #SBT protip: until Scala 2.10 final is released, you should declare your deps followed by 'cross CrossVersion.full' https://groups.google.com/d/msg/reactivemongo/yebmumHfp4E/LvFsrkHDMSoJ … 

Would changing `isStable` as follows buy us some time?

``` scala
def isStable(v: String): Boolean = !v.contains(".0-")
```

**Comment by jsuereth** [comment]
That may be enough to get us over a hump, yes.  I'm not positive though.

**Comment by jsuereth** [comment]
Yeah, thinking a bit more, we should also just avoid `-SNAPSHOT` completely in `isStable`

**Comment by eed3si9n** [comment]
Nightlies? If yes, how does one distinguish it with hotfix release `2.10.0-1`?

**Comment by jsuereth** [comment]
2.10.0-1 is stable, I was susggesting 

``` scala
def isStable(v: String): Boolean = !v.contains(".0-SNAPSHOT")
```

WHICH is only true if we're not looking at OSGi versions.  It's still a mostly broken method.

**Comment by eed3si9n** [comment]
ok. To clarify, I am going to call it "binary compatible with 2.10 API"

### bincompat with 2.10 API
- `2.10.0`
- `2.10.0-1`
- `2.10.1`
- `2.10.1-M1`
- `2.10.1-RC1`
- `2.10.1-SNAPSHOT`

### not bincompat with 2.10 API
- `2.10.0-M1`
- `2.10.0-RC1`
- `2.10.0-SNAPSHOT`

``` scala
def isAPICompatible(v: String): Boolean =
  !v.contains(".0-M") && !v.contains(".0-RC") && !v.contains(".0-SNAPSHOT")
```

**Comment by jsuereth** [comment]
What about a version like: `2.10.0-20121202-140434-70a24f907b` ?   This is the one we promote to OSGI, and OSGi assumes compatibility here.   If you're running off "named" snapshots, this is the verison you'd have.

The above is _not_ stable, while some later date would be.    Basically, what I'm saying is that unless the binary version is _encoded in the maven pom_, then there's absolutely no way to create an `isAPICompatible` function.   It just can't be done.

What you have is a decent 60% solution, but for a lot of my uses will fall over miserably, like the current algorithm.

So we're left with a question -  Do we improve what we have in an imperfect way, or do we attempt to push a real solution that drills all the way down to the dependency manager?

Knowing Mark, I assume he's pushing for the later.   Perhaps we should do both?

**Comment by eed3si9n** [comment]
Yea. That's what I meant by "Nightlies? If yes, how does one distinguish it with hotfix release `2.10.0-1`?"

In the short term we should fix the default broken state for 2.10.0 RCs. I doubt anyone would publish a library using a nightly build, so I think it's ok for them be recognized 0.12 compat. Meanwhile more people are publishing libs against RCs.

> Basically, what I'm saying is that unless the binary version is _encoded in the maven pom_, then there's absolutely no way to create an `isAPICompatible` function.

That will be a nice solution if you can convince Mark and the person who does jar publishing at Typesafe to add a parameter for `binaryVersion` starting 2.10.0 or 2.11.0. I can never remember the exact capability differences and issues with ivy.xml vs pom, but found my own post to ML [0.12 cross building](https://groups.google.com/forum/#!msg/simple-build-tool/6PdFaLaj4kM/vapUMxiURO0J):

> I did a bit of experiments to find out that translated ivy.xml sucks, but discovered a workaround too.
> For plugins, scalaVersion is already part of the [pom <properties> tag](https://oss.sonatype.org/content/repositories/public/com/eed3si9n/sbt-assembly_2.9.1_0.11.2/0.7.2/sbt-assembly-0.7.2.pom) and [Ivy's extra attributes on <info> tag](http://scalasbt.artifactoryonline.com/scalasbt/simple/sbt-plugin-releases/com.eed3si9n/sbt-assembly/scala_2.9.1/sbt_0.11.2/0.7.3/ivys/ivy.xml).
> Could you put these in for non plugins too?
> ...
> Within the same directory is ivy-%version.xml.original, which is POM.xml downloaded by Ivy when ivy.xml is not there.

**Comment by n8han** [comment]
> In the short term we should fix the default broken state for 2.10.0 RCs. I doubt anyone would publish a library using a nightly build, so I think it's ok for them be recognized 0.12 compat. Meanwhile more people are publishing libs against RCs.

Yeah, I ran into this just this week. I don't even want to deal with RCs (perhaps infamously) but library users do and they've persuaded me to publish releases. Previously I was using the default setting, because I figured I would have a release or three before 2.10 was actually released so it wouldn't matter too much.

But as we're getting closer, I don't want to publish releases that are incorrectly labeled as binary compatible with 2.10, so I had to figure out how to turn off the new CrossVersion.binary functionality. It _does the wrong thing_ for every 2.10 release to date. By only considering whether a release number is greater than 2.10, it is failing in only purpose -- stripping down the version tag when it is safe to do so.

I think that Eugene's proposed solution is good and should be merged, but if the argument is that we can't reliably determine weather a version string can be safely truncated, that's an argument for going back to CrossVersion.full by default. Surely the answer is not to err on the side of unsafety and leave it to library authors and users to clean up the mess?

**Comment by harrah** [comment]
The options are:
- 99% accuracy, 99% convenience: `_2.10`
- 100% accuracy, 5% convenience: `_2.10.0`
- 5% accuracy, ?% convenience: `<nothing>`

There is also the smarter selection of `_2.10` or `_2.10.0` as proposed by Eugene Y. above.  In theory that would be the best of both approaches, but the likelihood of making a bigger mess is high in my opinion.  This mixed approach was similar to the original implementation in 0.12.0 milestones, but user feedback on the mailing list was that it was not worth it.  It also had the issues outlined above.

I would love to be able to provide a 100% accurate solution without sacrificing convenience, but I don't think it is realistic to build that on top of the pom.xml format.  I experimented with a more accurate and more flexible approach when I tried to make it possible to cross-version sbt plugins against sbt and Scala and encode that in the pom.  Unfortunately, that ran afoul of repository manager verification, so it isn't practical and it wouldn't work for Maven users anyway.

As you probably remember, cross versioning came about because 2.7.x versions were grossly incompatible.  It wasn't a 1% chance of incompatibility, but a 99% chance of incompatibility.  You just couldn't use a library for Scala 2.7.0 with the 2.7.5 compiler because the compiler generated different bytecode or had a different pickling format or whatever.  Cross-versioning wasn't explicitly trying to make it 100% safe.  We were just trying to make it possible to work with multiple 2.7.x versions at all.

Now, we have two situations where the scheme is inaccurate: pre-releases of .0 versions and forward incompatibility, such as when a method is added in .1, for example.  This means that the inaccuracy is strongly present now, but as soon as 2.10.0 is out, it is reduced to the forward compatibility problem.  The full major.minor.micro scheme is mostly burden and little benefit at that point.  Certainly it will happen that people will run into incompatibilities with the major.minor scheme.  It would be great if we could prevent that.

In my opinion, cross-versioning should not be the solution to prevent every possible incompatibility.  It should be a tool to work with major incompatibilities.

**Comment by eed3si9n** [comment]
> This mixed approach was similar to the original implementation in 0.12.0 milestones, but user feedback on the mailing list was that it was not worth it.

I think there were initial confusion about what binary version is for sbt plugins, which may have lead to "not worth it" reactions. But after we've had some time to understand what it is, I think people generally like the idea of taking advantage of the binary compatibilities for Scala libs and sbt plugins.

Statistically, using `foo_2.10` to publish for 4 RCs and 1 final jar has been:
- 20% accuracy, ?% convenience: `_2.10`

The library authors are working around this, and even lead to Adriaan recommending to use full for 2.10.0 by mistake. The following is not perfect, but it's better than using `CrossVersion.binary` all the time.

``` scala
def isAPICompatible(v: String): Boolean =
  !v.contains(".0-M") && !v.contains(".0-RC") && !v.contains(".0-SNAPSHOT")
```

**Comment by viktorklang** [comment]
TL;DR

IMHO it could easily be encoded in the metadata for the scala-library.jar (what the default scalaBinaryVersion should be). It is not in the interest of the people depending on the Scala Library to have to change their build to get the correct behavior between RC cycle and final-cycle – it is also horribly error prone (as witnessed).

You should only ever have to tweak the scalaBinaryVersion if you know what you're doing and you know why you're doing it.

Cheers,
√

**Comment by gkossakowski** [comment]
@viktorklang: specific proposal are welcome. Remember that Scala library releases as Maven artifact with metadata encoded in `pom.xml`. Also, what exactly would like to encode in that metadata?

**Comment by viktorklang** [comment]
@gkossakowski Just add a scalaBinaryVersion here: https://github.com/scala/scala/blob/v2.9.2/src/library/scala/util/Properties.scala#L70

**Comment by gkossakowski** [comment]
@viktorklang: it looks to me like you'd have problem with different stages mixed up. The `scalaBinaryVersion` [is a Setting](https://github.com/harrah/xsbt/blob/0.13/main/src/main/scala/sbt/Keys.scala#L138) and not a Task so it's evaluated when project is loaded. I believe that managed dependency resolution happens _after_ all Settings are computed. It means that scala-library would be resolved after you compute value for `scalaBinaryVersion`.

Maybe `scalaBinaryVersion` can be turned into settings. I don't know that and I'll let @harrah comment on that point.

**Comment by harrah** [comment]
@viktorklang  As @gkossakowski suggests, you'd have to resolve scala-library before all other dependencies.  Having just cleaned up how the Scala instance is retrieved and handled, this is a complication that exceeds the budget.  The binary version needs to be known before resolution in the Maven model, unfortunately.

@eed3si9n It happened that the 0.12.0 RCs were binary compatible, so people preferred not having to republish plugins.  Binary compatibility isn't guaranteed for RCs, though.  For Scala, it has happened that the early RCs were not compatible.  If 2.11.0-RC1 is a more plausible release candidate, further RCs are more likely to be binary compatible.  Of course the empirical accuracy has been low so far- the inaccuracy is known to be in the prerelease stage.  The rest of the 2.10.x series will not have a problem and it will have 100x the users.

Anyway, you are making the right arguments, and if there is a change, it would be along the lines of what you propose.  The problem is that these kinds of things cannot easily be changed again.  The common case that affects the most users, stable releases, is correctly handled now and we can only make it worse.

Here is the best shot at something changing.  Make the `isAPICompatible` method properly parse the version into `X.Y.Z-*`.  If `*` is an integer or `-*` is not present, assume it is a release and compatible.  Otherwise, including if the version is not in the right form, it is not compatible.  We don't need to handle OSGi versions as far as I know.  Implement this, update the pull request, send an email to the mailing list, and link to this pull request.  I'm not in favor of this for the reasons above, but I know you'll maintain this and deal with ongoing issues.  Ideally, I'd like to see some kind of consensus from users who understand the issue.

**Comment by gkossakowski** [comment]
@eed3si9n, @harrah: I moved our discussion to mailing list and I outlined rather detailed proposal. I'd love to hear your feedback: https://groups.google.com/d/topic/simple-build-tool/mR8BAX_8kKI/discussion

**Comment by eed3si9n** [comment]
Added another commit, and rebased to the latest 0.13. Further discussion should be moved to the mailing list.

**Comment by harrah** [comment]
1. CrossVersion is public and needs to preserve source compatibility
2. Some indentation is messed up
3. I don't think Defaults needs to be changed.  The patch already redefines what CrossVersion.binary means.  When isScalaApiCompatible is false, binary is the same as full.

**Comment by eed3si9n** [comment]
@harrah 
1. I refactored it so it's less disrupting to the existing `CrossVersion`. I am putting `@deprecated` to `isStable` and `selectVersion` though.
2. Fixed spaces to tabs.
3. Rolled back Defaults.

```
> ++ 2.10.0-RC2
Setting version to 2.10.0-RC2
[info] Set current project to hello (in build file:/Users/foo/bar/)
> cross-version
[info] Binary
> scala-binary-version
[info] 2.10.0-RC2
```

**Comment by harrah** [comment]
Ok, looks good.  I merged it as is, but two more things: should binaryVersion be deprecated? and can you add tests?  A recent commit has tests for somewhat similar code and might be helpful: 824000009d11fb3fc0a2e491dc8c252c7bd85837.

**Comment by eed3si9n** [comment]
@harrah Deprecated `binaryVersion`, and added specs tests.

**Comment by harrah** [comment]
Good.  Should probably test things like 0.12.1-SNAPSHOT and -RC, right?

**Comment by eed3si9n** [comment]
@harrah added tests on non-final, but API compat versions.

**Comment by harrah** [comment]
Merged, thanks for all of your work on this.
